packages/frontend/core/src/modules/docs-search/entities/docs-indexer.ts

```diff
@@ -0,0 +1,299 @@
+            console.error('Error enqueueing job', err);
+          });
+      }
+    });
+  }
+
+  async execJob(jobs: Job<IndexerJobPayload>[], _signal: AbortSignal) {
+    if (jobs.length === 0) {
+      return;
+    }
+
+    // jobs should have the same docId, so we just pick the first one
+    const docId = jobs[0].payload.docId;
+
+    logger.debug('Start crawling job for docId:', docId);
+
+    if (docId) {
+      if (docId === this.workspaceId) {
+        await this.crawlingRootDocData();
+      } else {
+        await this.crawlingDocData(docId);
+      }
+    }
+  }
+
+  startCrawling() {
+    this.runner.start();
+    this.jobQueue
+      .enqueue([
+        {
+          batchKey: this.workspaceId,
+          payload: { docId: this.workspaceId },
+        },
+      ])
+      .catch(err => {
+        console.error('Error enqueueing job', err);
+      });
+  }
+
+  async crawlingDocData(docId: string) {
+    const rootDocBuffer =
+      await this.workspaceEngine.doc.storage.loadDocFromLocal(this.workspaceId);
+
+    if (!rootDocBuffer) {
+      return;
+    }
+
+    const yRootDoc = new YDoc();
+    applyUpdate(yRootDoc, rootDocBuffer);
+
+    const docStoragePossibleIds = Array.from(yRootDoc.getSubdocs())
+      .map(doc => doc.guid)
+      .filter(id => id.endsWith(docId));
+
+    let docBuffer;
+
+    for (const id of docStoragePossibleIds) {
+      docBuffer = await this.workspaceEngine.doc.storage.loadDocFromLocal(id);
+    }
+
+    if (!docBuffer) {
+      return;
+    }
+
+    const ydoc = new YDoc();
+
+    applyUpdate(ydoc, docBuffer);
+
+    let docExists: boolean | null = null;
+
+    (
+      yRootDoc.getMap('meta').get('pages') as YArray<YMap<any>> | undefined
+    )?.forEach(page => {
+      if (page.get('id') === docId) {
+        docExists = !(page.get('trash') ?? false);
+      }
+    });
+
+    if (!docExists) {
+      const indexWriter = await this.docIndex.write();
+      indexWriter.delete(docId);
+      await indexWriter.commit();
+
+      const blockIndexWriter = await this.blockIndex.write();
+      const oldBlocks = await blockIndexWriter.search(
+        {
+          type: 'match',
+          field: 'docId',
+          match: docId,
+        },
+        {
+          pagination: {
+            limit: Number.MAX_SAFE_INTEGER,
+          },
+        }
+      );
+      for (const block of oldBlocks.nodes) {
+        blockIndexWriter.delete(block.id);
+      }
+      await blockIndexWriter.commit();
+    } else {
+      const blocks = ydoc.getMap<any>('blocks');
+
+      if (blocks.size === 0) {
+        return;
+      }
+
+      let docTitle = '';
+
+      const blockDocuments: Document<typeof blockIndexSchema>[] = [];
+
+      for (const block of blocks.values()) {
+        const flavour = block.get('sys:flavour')?.toString();
+        const blockId = block.get('sys:id')?.toString();
+
+        if (!flavour || !blockId) {
+          continue;
+        }
+
+        if (flavour === 'affine:page') {
```

Comment:
I think eventually every block can define the crawler for itself.

Reply:
Need to be define in blocksuite